File: 리트코드/0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return dfs(0, 0, inorder.length - 1, preorder, inorder);

    }

    public TreeNode dfs(int preIndex, int inStart, int inEnd, int[] preorder, int[] inorder) {

        if (inStart > inEnd) {
            return null;
        }

        int inIndex = 0;
        for (int i = inStart; i <= inEnd; i++) {
            if (inorder[i] == preorder[preIndex]) {
                inIndex = i;
                break;
            }
        }

        TreeNode treeNode = new TreeNode(inorder[inIndex]);

        treeNode.left = dfs(preIndex + 1, inStart, inIndex - 1, preorder, inorder);
        treeNode.right = dfs(preIndex + 1 + (inIndex - inStart), inIndex + 1, inEnd, preorder, inorder);

        return treeNode;
    }
}

File: 리트코드/0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun buildTree(preorder: IntArray, inorder: IntArray): TreeNode? {
        fun dfs(preIndex: Int, start: Int, end: Int): TreeNode? {

            if (start > end) return null

            var inIndex: Int = -1;
            for (i in start..end) {
                if (inorder[i] == preorder[preIndex]) {
                    inIndex = i
                    break
                }
            }

            val node: TreeNode = TreeNode(preorder[preIndex])

            node.left = dfs(preIndex + 1, start, inIndex - 1)
            node.right = dfs(preIndex + 1 + (inIndex - start), inIndex + 1, end)
            
            return node
        }

        return dfs(0, 0, inorder.size - 1)
        
    }
}